notificationworkerlambda/src/main/scala/com/gu/notifications/worker/Harvester.scala

```diff
@@ -0,0 +1,94 @@
+package com.gu.notifications.worker
+
+import _root_.models.{Android, Newsstand, Platform, ShardedNotification, iOS}
+import cats.effect.{ContextShift, IO, Timer}
+import com.amazonaws.services.lambda.runtime.Context
+import com.amazonaws.services.lambda.runtime.events.SQSEvent
+import com.gu.notifications.worker.tokens.{ChunkedTokens, SqsDeliveryService, SqsDeliveryServiceImpl, TokenService, TokenServiceImpl}
+import com.gu.notifications.worker.utils.{Cloudwatch, CloudwatchImpl, Logging, NotificationParser}
+import db.{DatabaseConfig, RegistrationService}
+import doobie.util.transactor.Transactor
+import fs2.{Sink, Stream}
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, ExecutionContextExecutor}
+
+
+trait HarvesterRequestHandler extends Logging {
+  implicit val ec: ExecutionContextExecutor = ExecutionContext.global
+  implicit val ioContextShift: ContextShift[IO] = IO.contextShift(ec)
+  implicit val timer: Timer[IO] = IO.timer(ec)
+  implicit val logger: Logger = LoggerFactory.getLogger(this.getClass)
+  val env = Env()
+  val apnsDeliveryService: SqsDeliveryService[IO]
+  val firebaseDeliveryService: SqsDeliveryService[IO]
+  val tokenService: TokenService[IO]
+  val cloudwatch: Cloudwatch
+  val maxConcurrency: Int = 100
+  val logErrors: Sink[IO, Throwable] = throwables => {
+    throwables.map(throwable => logger.warn("Error queueing", throwable))
+  }
+
+  def sinkErrors(platform: Platform): Sink[IO, Throwable] = throwables => {
+    throwables.broadcastTo(logErrors, cloudwatch.sendFailures(env.stage, platform))
+  }
+
+  val firebaseSink: Sink[IO, ChunkedTokens] = {
+    val androidSinkErrors = sinkErrors(Android)
+    chunkedTokens => {
+      chunkedTokens.filter(_.platform == Android)
+        .map(firebaseDeliveryService.sending)
+        .parJoin(maxConcurrency)
+        .collect {
+          case Left(throwable) => throwable
+        }
+        .to(androidSinkErrors)
+    }
+  }
+
+  val apnsSink: Sink[IO, ChunkedTokens] = {
+    val iosSinkErrors = sinkErrors(iOS)
+    chunkedTokens => {
+      chunkedTokens.filter(chunkedTokens => chunkedTokens.platform == iOS || chunkedTokens.platform == Newsstand)
+        .map(apnsDeliveryService.sending)
+        .parJoin(maxConcurrency)
+        .collect {
+          case Left(throwable) => throwable
+        }
+        .to(iosSinkErrors)
+    }
+  }
+
+  def queueShardedNotification(shardNotifications: Stream[IO, ShardedNotification]): Stream[IO, Unit] = {
+    val chunkedTokenStream = for {
+      n <- shardNotifications
+      platform = n.platform
+      notificationLog = s"(notification: ${n.notification.id} ${n.range})"
+      _ = logger.info(s"Queuing notification $notificationLog...")
+      tokens <- tokenService.tokens(n.notification, n.range, platform).chunkN(1000)
```

Comment:
I think the chunking should happen in the firebase and the apns sink, otherwise we're going to have batches of about 500 tokens each